app/cli/commandlineparser.cpp

```diff
@@ -0,0 +1,363 @@
+        m_Choices[name] = choices;
+    }
+
+private:
+    QMap<QString, QStringList> m_Choices;
+};
+
+GlobalCommandLineParser::GlobalCommandLineParser()
+{
+}
+
+GlobalCommandLineParser::~GlobalCommandLineParser()
+{
+}
+
+GlobalCommandLineParser::ParseResult GlobalCommandLineParser::parse(const QStringList &args)
+{
+    CommandLineParser parser;
+    parser.setupCommonOptions();
+    parser.setApplicationDescription(
+        "\n"
+        "Starts Moonlight normally if no arguments are given.\n"
+        "\n"
+        "Available actions:\n"
+        "  stream          Start streaming a game\n"
+        "\n"
+        "See 'moonlight <action> --help' for help of specific action."
+    );
+    parser.addPositionalArgument("action", "Action to execute", "<action>");
+    parser.parse(args);
+    auto posArgs = parser.positionalArguments();
+    QString action = posArgs.isEmpty() ? QString() : posArgs.first().toLower();
+
+    if (action == "") {
+        // This method will not return and terminates the process if --version
+        // or --help is specified
+        parser.handleHelpAndVersionOptions();
+        parser.handleUnknownOptions();
+        return NormalStartRequested;
+    } else if (action == "stream") {
+        return StreamRequested;
+    } else {
+        parser.showError(QString("Invalid action: %1").arg(action));
+    }
+}
+
+StreamCommandLineParser::StreamCommandLineParser()
+{
+    m_WindowModeMap = {
+        {"fullscreen", StreamingPreferences::WM_FULLSCREEN},
+        {"windowed",   StreamingPreferences::WM_WINDOWED},
+        {"borderless", StreamingPreferences::WM_FULLSCREEN_DESKTOP},
+    };
+    m_AudioConfigMap = {
+        {"auto",     StreamingPreferences::AC_AUTO},
+        {"stereo",   StreamingPreferences::AC_FORCE_STEREO},
+        {"surround", StreamingPreferences::AC_FORCE_SURROUND},
+    };
+    m_VideoCodecMap = {
+        {"auto",  StreamingPreferences::VCC_AUTO},
+        {"H.264", StreamingPreferences::VCC_FORCE_H264},
+        {"HEVC",  StreamingPreferences::VCC_FORCE_HEVC},
+    };
+    m_VideoDecoderMap = {
+        {"auto",     StreamingPreferences::VDS_AUTO},
+        {"software", StreamingPreferences::VDS_FORCE_HARDWARE},
+        {"hardware", StreamingPreferences::VDS_FORCE_SOFTWARE},
+    };
+}
+
+StreamCommandLineParser::~StreamCommandLineParser()
+{
+}
+
+void StreamCommandLineParser::parse(const QStringList &args, StreamingPreferences *preferences)
+{
+    CommandLineParser parser;
+    parser.setupCommonOptions();
+    parser.setApplicationDescription(
+        "\n"
+        "Starts directly streaming a given game."
+    );
+    parser.addPositionalArgument("stream", "Start stream");
+
+    // Add other arguments and options
+    parser.addPositionalArgument("host", "Host computer name, uuid or address", "<host>");
+    parser.addPositionalArgument("game", "Game name", "<game>");
```

Comment:
let's put quotes around `<game>` to avoid confusion with users specifying games with spaces but not using quotes.

Reply:
Done.
Which one you think is better here; game, app, or application? I'm not sure which one I should use. Goes also for text in `setApplicationDescription()`.

Reply:
I think the text content in first argument `app` and third `"<app name>"` should remain same. So `app` and `"<app name>"` or `app` and third `"<app>"`. Otherwise there is a confusion between usage text and arguments list:
```
> moonlight stream --help
Usage: moonlight [options] stream <host> "<app name>"
...
Arguments:
...
  app                              App to stream
```

On the other hand I feel that having `app name` in the arguments list might be confusing, since white space has a meaning as a argument separator.

Thus I would keep the third parameter as `<app>`

Reply:
That sounds ok to me.